File: core/src/main/scala/eu/easyminer/rdf/data/Transformable.scala
package eu.easyminer.rdf.data

import scala.collection.TraversableView

/**
  * Created by Vaclav Zeman on 2. 2. 2018.
  */
trait Transformable[T, Coll] {

  protected def coll: TraversableView[T, Traversable[_]]

  protected def transform(col: TraversableView[T, Traversable[_]]): Coll

  def map(f: T => T): Coll = transform(coll.map(f))

  def filter(f: T => Boolean): Coll = transform(coll.filter(f))

  def slice(from: Int, until: Int): Coll = transform(coll.slice(from, until))

  def take(n: Int): Coll = transform(coll.take(n))

  def drop(n: Int): Coll = transform(coll.drop(n))

}

File: java/src/main/java/com/github/propi/rdfrules/java/data/Transformable.java
package com.github.propi.rdfrules.java.data;

import java.util.function.Predicate;
import java.util.function.UnaryOperator;

/**
 * Created by Vaclav Zeman on 2. 5. 2018.
 */
public interface Transformable<ST, JT, SColl, JColl> {

    com.github.propi.rdfrules.data.ops.Transformable<ST, SColl> asScala();

    JT asJavaItem(ST x);

    ST asScalaItem(JT x);

    JColl asJava(SColl scala);

    default JColl map(UnaryOperator<JT> f) {
        return asJava(asScala().map(x -> asScalaItem(f.apply(asJavaItem(x)))));
    }

    default JColl filter(Predicate<JT> f) {
        return asJava(asScala().filter(x -> f.test(asJavaItem(x))));
    }

    default JColl slice(int from, int until) {
        return asJava(asScala().slice(from, until));
    }

    default JColl take(int n) {
        return asJava(asScala().take(n));
    }

    default JColl drop(int n) {
        return asJava(asScala().drop(n));
    }

    default int size() {
        return asScala().size();
    }

}